pkg/cmd/server/start.go

```diff
@@ -273,365 +129,40 @@ func start(cfg *config, args []string) error {
-			k8sPublicAddr = cfg.KubernetesPublicAddr
-		} else if cfg.KubernetesAddr.Provided { // separate Kube, assume it is public
-			k8sPublicAddr = cfg.KubernetesAddr
-		}
-
-		// Derive the asset bind address by incrementing the master bind address port by 1
-		assetBindAddr := net.JoinHostPort(cfg.BindAddr.Host, strconv.Itoa(cfg.BindAddr.Port+1))
-		// Derive the asset public address by incrementing the master public address port by 1
-		assetPublicAddr := *masterPublicAddr.URL
-		assetPublicAddr.Host = net.JoinHostPort(masterPublicAddr.Host, strconv.Itoa(masterPublicAddr.Port+1))
-
-		// Build the list of valid redirect_uri prefixes for a login using the openshift-web-console client to redirect to
-		// TODO: allow configuring this
-		// TODO: remove hard-coding of development UI server
-		assetPublicAddresses := []string{assetPublicAddr.String(), "http://localhost:9000", "https://localhost:9000"}
-
-		// always include the all-in-one server's web console as an allowed CORS origin
-		// always include localhost as an allowed CORS origin
-		// always include master public address as an allowed CORS origin
-		for _, origin := range []string{assetPublicAddr.Host, masterPublicAddr.URL.Host, "localhost", "127.0.0.1"} {
-			// TODO: check if origin is already allowed
-			cfg.CORSAllowedOrigins = append(cfg.CORSAllowedOrigins, origin)
-		}
-
-		osmaster := &origin.MasterConfig{
-			TLS:                  cfg.BindAddr.URL.Scheme == "https",
-			MasterBindAddr:       cfg.BindAddr.URL.Host,
-			MasterAddr:           cfg.MasterAddr.URL.String(),
-			MasterPublicAddr:     masterPublicAddr.URL.String(),
-			AssetBindAddr:        assetBindAddr,
-			AssetPublicAddr:      assetPublicAddr.String(),
-			KubernetesAddr:       cfg.KubernetesAddr.URL.String(),
-			KubernetesPublicAddr: k8sPublicAddr.URL.String(),
-
-			LogoutURI: env("OPENSHIFT_LOGOUT_URI", ""),
-
-			CORSAllowedOrigins: cfg.CORSAllowedOrigins,
-
-			EtcdHelper: etcdHelper,
-
-			AdmissionControl:              admit.NewAlwaysAdmit(),
-			Authorizer:                    newAuthorizer(etcdHelper, masterAuthorizationNamespace),
-			AuthorizationAttributeBuilder: newAuthorizationAttributeBuilder(requestContextMapper),
-			MasterAuthorizationNamespace:  masterAuthorizationNamespace,
-			RequestContextMapper:          requestContextMapper,
-
-			UseLocalImages: useLocalImages,
-			ImageFor:       imageResolverFn,
-		}
-
-		if startKube {
-			// We're running against our own kubernetes server
-			osmaster.KubeClientConfig = kclient.Config{
-				Host:    cfg.MasterAddr.URL.String(),
-				Version: klatest.Version,
-			}
-		} else {
-			// We're running against another kubernetes server
-			osmaster.KubeClientConfig = *clientConfigFromKubeConfig(cfg)
-		}
+	}
 
-		// Build token auth for user's OAuth tokens
-		authenticators := []authenticator.Request{}
-		tokenAuthenticator, err := origin.GetEtcdTokenAuthenticator(etcdHelper)
+	if cfg.StartNode {
+		openshiftConfig, err := cfg.BuildOriginMasterConfig()
```

Comment:
If you run this on a standalone node, it'll result in generating unused server and client certs on the node... should we have a NodeConfig/BuildNodeConfig() that just gets what we need for a node?

* bind addr
* master addr
* kube client
* eventually, cert/key for starting node on https